### ISSUE TITLE
Unable to install / create config

### ISSUE THREAD

**Issue by dj3hac** [aclist/dztui#65]
I see another user has posted this same issue.

I am unable to install DZGUI as it is unable to determine my game path. I was able to configure this ~last year when I was running Pop_OS. I put in my username, and 2x API keys before running into "DayZ not found or not installed at the chosen path.". I do not get the option to choose an install path manually.

Nobara (Fedora) 37
Flatpak steam

Library where the game is installed: /mnt/SSD3/SteamLibrary

**Comment by aclist** [comment]
Thanks for the report and details, and for checking for open issues. I am 99% certain this is a Flatpak issue after checking the path discovery logic again. We just need to tighten up the path expansion so that it doesn't accidentally pick Flatpak symlink directories as the actual path. I pushed a hotfix into the stable branch, so please try installing again and seeing if the path is picked up.

If you still run into problems after the hotfix (version `3.2.6`): as this is a duplicate of #65, let's merge this into there while still giving you individualized support. For the sake of completeness, could you reply there with the additional information below? I will follow up after I get back to the desk.

- DZGUI version (assume latest?)
- DZGUI branch (stable/testing; assume stable?)
- Steam branch (stable or beta)
- Method of invocation (terminal, shortcut, or non-Steam game shortcut inside Steam client)

Thanks.

**Comment by dj3hac** [comment]
The hotfix did not change anything for me, unfortunately.

    DZGUI version: 3.2.6
    DZGUI branch: Stable
    Steam branch: Stable
    Method of invocation: curl -s "https://raw.githubusercontent.com/aclist/dztui/dzgui/install.sh" | bash

**Comment by aclist** [comment]
Can you post the output of this?

`find / -type d \( -path "/proc" -o -path "*/timeshift" -o -path "/tmp" -o -path "/usr" -o -path "/boot" -o -path "$HOME/.var" -o -path "/proc" -o -path "/root" -o -path "/sys" -o -path "/etc" -o -path "/var" -o -path "/lost+found" \) -prune -o -regex ".*/Steam/ubuntu12_32$" -print -quit 2>/dev/null | sed 's@/ubuntu12_32@@'`

I'll be back in 20-30 minutes.

**Comment by dj3hac** [comment]
/home/dj3hac/.local/share/Steam

**Comment by aclist** [comment]
OK, try this now:

`python3 $HOME/.local/share/dzgui/helpers/vdf2json.py -i $HOME/.local/share/Steam/steamapps/libraryfolders.vdf | jq -r '.libraryfolders[]|select(.apps|has("221100")).path'`

**Comment by aclist** [comment]
Also, have you installed Steam in multiple locations? Is it also installed somewhere other than `/home/dj3hac/.local/share/Steam`?

**Comment by aclist** [comment]
I pushed an update that fixed the problem with getting caught in a loop of "DayZ not found" and being unable to manually specify a path. This doesn't directly address your underlying problem, just offers an escape hatch. But I am guessing that you had concurrent Steam directories somewhere and DZGUI defaulted to the first one it found.

**Comment by dj3hac** [comment]
Hey, I'm at work right now, so I can't test this. I will later though! One thing I forgot to mention is that ORIGINALLY I had the RPM version of steam installed, but it was not playing nicely with counterstrike so I switched to the flatpak version and have been using that ever since. I had un-installed the RPM version a long time ago, this version did not have dayz installed at any point. On May 10, 2023 14:34, aclist ***@***.***> wrote:
I pushed an update that fixed the problem with getting caught in a loop of "DayZ not found" and being unable to manually specify a path. This doesn't directly address your underlying problem, just offers an escape hatch. But I am guessing that you had concurrent Steam directories somewhere and DZGUI defaulted to the first one it found.

**Comment by dj3hac** [comment]
> OK, try this now:
> 
> `python3 $HOME/.local/share/dzgui/helpers/vdf2json.py -i $HOME/.local/share/Steam/steamapps/libraryfolders.vdf | jq -r '.libraryfolders[]|select(.apps|has("221100")).path'`

Nothing seems to have happened with this.

![image](https://github.com/aclist/dztui/assets/38978989/047d6f19-2387-411a-9588-8ccd883c0567)

**Comment by dj3hac** [comment]
> I pushed an update that fixed the problem with getting caught in a loop of "DayZ not found" and being unable to manually specify a path. This doesn't directly address your underlying problem, just offers an escape hatch. But I am guessing that you had concurrent Steam directories somewhere and DZGUI defaulted to the first one it found.

The newest update did not work either, pointing to my steam library gave me this error: 
![image](https://github.com/aclist/dztui/assets/38978989/11d53974-4741-4bf5-8397-fb393203d9b7)

Directory supplied: /mnt/SSD3/SteamLibrary

**Comment by aclist** [comment]
I see the issue. It looks like Steam is installed at `$HOME/.local/share/Steam` and your "library" resides at `/mnt...`. The library isn't where the canonical Steam installation resides, it's just where the game files are. DZGUI expects the location to Steam itself. You don't need to supply the location of games, just the location to Steam. Hope this makes sense. Example:

I have games installed to `/media/nvme`. DZGUI is set to `$HOME/.local/share/Steam`. Steam reports that `/media/nvme` is one of its library locations. DZGUI picks up the rest from here. If Steam had been installed to, e.g., `$HOME/Steam`, I would supply `$HOME/Steam` in the file picker.

However, it looks like the Steam location found at `$HOME/.local/share/Steam` isn't reporting that `/mnt...` is one of the directories available to it. This causes me to conclude that the installation of Steam at `$HOME/.local/share/Steam` is not the one you installed DayZ to.

Since you said you had installed Steam a few different times, can you locate where Steam is actually installed now?

Let me know if you need help formulating a command to scan the system and find Steam.

**Comment by aclist** [comment]
Given that you said Flatpak, it looks like you should be specifying `$HOME/.var/app/com.valvesoftware.Steam/data/Steam`

**Comment by dj3hac** [comment]
> Given that you said Flatpak, it looks like you should be specifying `$HOME/.var/app/com.valvesoftware.Steam/data/Steam`

You are correct! This is where steam is installed however when pointing to that directory, as well as a few directories below it for good measure I run into the same isse: 

![image](https://github.com/aclist/dztui/assets/38978989/9af3747e-aff0-471f-a36f-4d486830fc3c)

**Comment by aclist** [comment]
Try: 

`python3 $HOME/.local/share/dzgui/helpers/vdf2json.py -i $HOME/.var/app/com.valvesoftware.Steam/data/steam/steamapps/libraryfolders.vdf | jq '.libraryfolders[]|select(.apps|has("221100")).path'

**Comment by dj3hac** [comment]
Nothing

![image](https://github.com/aclist/dztui/assets/38978989/b015f405-a6e9-410b-b783-ee1d5607572d)

**Comment by aclist** [comment]
There was a typo, remove the leading backtick.

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/04104c48-1736-474a-8603-41dbbed4aa93)

It says the file does not exist, however you can see in my file browser that it is existing in the expected location..

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/98e0689d-2a62-479c-a351-437437f3e5fa)

Capital S in "Steam"

**Comment by aclist** [comment]
Ouch, good find. OK, seems we are almost there, just thinking about the ramifications of this.

**Comment by aclist** [comment]
It is clear that DayZ is installed there as reported by Steam itself. I think this has to be some form of user error when navigating the file picker. Are you going too deep into the directory tree? Your selection target should go no deeper than `Steam`. You should be navigating into the folder `$HOME/.var/app/com.valvesoftware.Steam/data/`, highlighting that, and then selecting `Steam` as the final directory and hitting the OK button. If you navigate into `Steam` and open something else, you are already too far.

I'm sorry this isn't clear enough from the menus, but it is explained in the documentation. I could look into prompting the user at the beginning to choose between Flatpak or not and then search based on that, but I think the problem is that people install, uninstall, and wind up managing a bunch of different Steam versions on their machines without pruning them, so I'm not sure if it should be the responsibility of Steam to try to guess which is the right one.

**Comment by dj3hac** [comment]
1) Download fresh install.sh
![image](https://github.com/aclist/dztui/assets/38978989/a1db63c4-6ffb-4023-adfe-539719a1bfc7)

2) Begin setup.
![image](https://github.com/aclist/dztui/assets/38978989/ab68c321-8ea8-4246-9579-c57e21dc15f9)

3) User / API keys.
![image](https://github.com/aclist/dztui/assets/38978989/2c05e6e5-0f6b-4036-a793-f3fe1e14809b)

4) Where's Waldo?
![image](https://github.com/aclist/dztui/assets/38978989/61cf4928-866f-491b-af19-7d39d45659b8)

5) Waldo lives here.
![image](https://github.com/aclist/dztui/assets/38978989/e0ade743-4a12-403b-a83e-d88edc451860)

6) Cries in open source.
![image](https://github.com/aclist/dztui/assets/38978989/abca3170-819d-4de3-8342-575507cc84a3)

**Comment by aclist** [comment]
Judging from the last image, you went too deep. You're already inside the `Steam` directory by then, so your selection when you hit `OK` is going to be some file inside of there. You need to be in `data` and highlight the `Steam` folder before hitting OK. You can simulate this by using `zenity --file-selection --directory` from the terminal, navigating, hitting OK, and seeing what path is printed to stdout.

**Comment by dj3hac** [comment]
> Judging from the last image, you went too deep. You're already inside the `Steam` directory by then, so your selection when you hit `OK` is going to be some file inside of there. You need to be in `data` and highlight the `Steam` folder before hitting OK. You can simulate this by using `zenity --file-selection --directory` from the terminal, navigating, hitting OK, and seeing what path is printed to stdout.

I figured you would say this, I did not have anything selected inside the directory. I tried both, I went back and did not open the steam dir and had the same issue. I have tried every folder from here all the way back to my home directory. 

Please remember that I'm a former user of DZGUI, this works fine on pop OS, it does not work on Fedora. Do you want to remote into my system and see it not work for yourself?

**Comment by aclist** [comment]
OK, I am thinking. Thus far I have been unable to reproduce it. I'm struggling to see where the bug is, here, but of course it could be a possibility. zenity is an unmaintained library and contains some idiosyncrasies. Just now I observed that trying to type the path and hit Enter caused it to fail prematurely, whereas manually selecting directories with the mouse worked.

I don't see anything as yet that indicates a distro-specific problem (I am not aware of many distro-specific idiosyncrasies affecting DZGUI, either), aside from the outside possibility of some permission issue, but it looks like everything can be probed as expected based on the other tests. It's not that I don't believe you, I just have to methodically try every possibility here to isolate false positives, as tedious as it may be. Of course, there might be something obvious right in front of my face that I am missing.

Here is a simplified unit test that breaks the process down into something smaller without the overhead of DZGUI. Make this executable and run it.

```
#!/bin/bash

sd_install_path="$HOME/.local/share/dzgui"
helpers_path="$sd_install_path/helpers"

if [[ -f $helpers_path/vdf2json.py ]]; then
	echo "FOUND HELPER FILE"
else
	echo "FAILED TO FIND HELPER"
	exit 1
fi

find_library_folder(){
	python3 $helpers_path/vdf2json.py -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path'
}

file_picker(){
	echo "OPENING FILE PICKER"
	local path=$(zenity --file-selection --directory 2>/dev/null)
	if [[ -z "$path" ]]; then
		echo "NO PATH SELECTED"
		return
	fi
	echo "USER SELECTED: $path"
	printf "Install path is "
	find_library_folder "$path"
}

file_picker
```

**Comment by dj3hac** [comment]
https://github.com/aclist/dztui/assets/38978989/31e30d60-f9e9-4620-b978-2f32eb8635a1

It seems to crash when I select the steam folder. Am I correct in assuming that what I'm supposed to be doing?

**Comment by aclist** [comment]
Yes, that's very interesting. Does this happen when invoking zenity by itself? Let's write the errors to a file and see what it contains:

`zenity --file-selection --directory 2>$HOME/zenitylog`

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/6e1910ac-65bc-415b-9ea8-af1b069701f6)

This seems to have executed without error

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/4ab07493-086e-4716-a478-b28a9ae0e4a3)

**Comment by aclist** [comment]
Thanks for the additional details. I have a few theories so far. Could be a GTK bug that is silently crashing the file picker in the background and never passing the expected path to DZGUI. Could be that zenity has a bug that doesn't behave well when redirecting stderr.

Let's try this version:

```
#!/bin/bash

sd_install_path="$HOME/.local/share/dzgui"
helpers_path="$sd_install_path/helpers"

if [[ -f $helpers_path/vdf2json.py ]]; then
	echo "FOUND HELPER FILE"
else
	echo "FAILED TO FIND HELPER"
	exit 1
fi

find_library_folder(){
	python3 $helpers_path/vdf2json.py -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path'
}

file_picker(){
	echo "OPENING FILE PICKER"
	local path=$(zenity --file-selection --directory)
	if [[ -z "$path" ]]; then
		echo "NO PATH SELECTED"
		return
	fi
	echo "USER SELECTED: $path"
	printf "Install path is "
	find_library_folder "$path"
}

file_picker
```

And this version:

```
#!/bin/bash

path=$(zenity --file-selection --directory 2>$HOME/zenitylog2)
echo $path
```

**Comment by aclist** [comment]
> Not sure what "0" means, but since it's the only "0" it's probably not a good thing.

This won't block discovery of the drive path itself. It's failing when parsing the library folders because it's never getting the path returned from the file selection dialog, apparently, and thus can't efven find the library folders file.

**Comment by aclist** [comment]
OK, I was able to reproduce this finally. Seems to be a zenity bug relating to spawning a process in a subshell. The value isn't getting returned as it should. Please stand by, I will come up with a hotfix. Everything else is working as intended.

**Comment by dj3hac** [comment]
Hey, that's great! Hopefully you've found the root of the issue. Is this something to do with my configuration, or is it a distro thing? I think I saw somewhere you were looking for Fedora testers, so I imagine Fedora hasn't been tested too much.

![image](https://github.com/aclist/dztui/assets/38978989/769a1b8a-ca5c-41d0-b996-2ffb92978b1d)

**Comment by aclist** [comment]
I can't say for certain without ruling out more possibilities, but I think it is an undocumented behavior in the zenity package itself or how it interacts with GTK. For now I am looking to prevent recurrence before hardening it further.

I think DZGUI is mostly feature complete at this point, but I'm always looking to harden it against obscure bugs. I will also need testers for my forthcoming other application.

Hopefully the below works. I also created a Flatpak directory and put the files in the same structure as yours to test it. Provided we can get the file picker to stop crashing, the rest should be fine.

```
#!/bin/bash

sd_install_path="$HOME/.local/share/dzgui"
helpers_path="$sd_install_path/helpers"

if [[ -f $helpers_path/vdf2json.py ]]; then
	echo "FOUND HELPER FILE"
else
	echo "FAILED TO FIND HELPER"
	exit 1
fi

find_library_folder(){
	python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path'
}

file_picker(){
	echo "OPENING FILE PICKER"
	local t="/tmp/dzguipath"
	zenity --file-selection --directory > $t
	if [[ ! -s "$t" ]]; then
		echo "NO PATH SELECTED"
		rm $t
		return
	fi
	path=$(cat $t)
	echo "USER SELECTED: $path"
	printf "Install path is "
	find_library_folder "$path"
	rm $t
}

file_picker
```

**Comment by aclist** [comment]
Wait, there's something else I just noticed in your video. It looks like the process was pinning to that terminal window and the whole thing exited when zenity failed. Did you by chance run the script by double clicking on it through the file explorer, or did you invoke it manually in the terminal, e.g., `./dztest.sh` or `sh dztest.sh`? Just want to make sure, as spawning scripts through system file navigators can be the cause of numerous problems.

**Comment by dj3hac** [comment]
> Did you by chance run the script by double clicking on it through the file explorer

Shh.. let's not talk about that.. 

![image](https://github.com/aclist/dztui/assets/38978989/c0ded3ff-6596-462d-8a8c-7b389ceb7f15)

Running the 1st script again, properly, does find my install path on the correct drive.

**Comment by dj3hac** [comment]
Running the install script by invoking it through the terminal is still unable to find dayz.

The 2nd and 3rd script appear to work as well.

![image](https://github.com/aclist/dztui/assets/38978989/e6b464ac-a455-4d1a-ad2e-9e3d39b8df0b)

**Comment by aclist** [comment]
OK. The problem with the plugins that let you launch scripts via right-click context menus is that they are designed mostly with very simple scripts in mind that just do one thing and exit. If the script is more complex and needs to use subshells or set a certain environment, it's possible it can get swallowed into the environment the file explorer spawned, typically causing unpredictable and difficult to diagnose problems. 

I've made some slight changes to the DZGUI install script on the testing branch. You can just grab the file directly from this link, `chmod +x` it, and run it from the terminal `./dzgui.sh`.

https://raw.githubusercontent.com/aclist/dztui/testing/dzgui.sh

**Comment by aclist** [comment]
Sorry, disregard the above. I had unmerged changes between the two branches due to a host of tickets that got submitted these last few days. Please use this instead:

` curl "https://raw.githubusercontent.com/aclist/dztui/main/install.sh" | bash`

**Comment by dj3hac** [comment]
> OK. The problem with the plugins that let you launch scripts via right-click context menus is that they are designed mostly with very simple scripts in mind that just do one thing and exit. If the script is more complex and needs to use subshells or set a certain environment, it's possible it can get swallowed into the environment the file explorer spawned, typically causing unpredictable and difficult to diagnose problems.

This makes a lot of sense.

> I've made some slight changes to the DZGUI install script on the testing branch. You can just grab the file directly from this link, `chmod +x` it, and run it from the terminal `./dzgui.sh`.
> 
> https://raw.githubusercontent.com/aclist/dztui/testing/dzgui.sh

This also did not work. I did try running it as superuser and that completed the install, and I can launch DZGUI as su (not a regular user, it asks me to configure again). I can't test if this works because now my steam client is bricked.

![image](https://github.com/aclist/dztui/assets/38978989/e7c742db-daa1-4e4f-8bfb-67ed6f19e8d3)

**Comment by dj3hac** [comment]
> Sorry, disregard the above. I had unmerged changes between the two branches due to a host of tickets that got submitted these last few days. Please use this instead:
> 
> ` curl "https://raw.githubusercontent.com/aclist/dztui/main/install.sh" | bash`

Too late 
![image](https://github.com/aclist/dztui/assets/38978989/8e29be1a-a472-49d6-a874-f220df4b38c8)

**Comment by dj3hac** [comment]
Okay, I've fixed steam:

/home/dj3hac/.local/share/flatpak/overrides/com.valvesoftware.Steam

Removed "host" & "home" entry.

**Comment by aclist** [comment]
Oops. OK, so is the mainline DZGUI install working?

**Comment by aclist** [comment]
About the superuser thing, there is some old code kicking around on the testing branch that needs to be refactored to backport changes from the stable branch, because that sudo escalation prompt should be dropped.

**Comment by dj3hac** [comment]
> curl "https://raw.githubusercontent.com/aclist/dztui/main/install.sh" | bash

Just tried this latest rendition, unfortunately not. Same error, unable to locate install.

It also says im on version 3.2.10 and need to upgrade to 3.2.9 it seems to not actually update as it then tells me I have the same mis-match after restarting dzgui

![image](https://github.com/aclist/dztui/assets/38978989/9b8eae9e-6779-4fcb-8eb1-0e5afc72d5d4)

**Comment by aclist** [comment]
Version mismatch: fixed and updated all branches to the latest code.

DayZ not found bug: expanded version of initial tests follows. Does this work? This is the minimum reproducible example.

```
#!/bin/bash

sd_install_path="$HOME/.local/share/dzgui"
helpers_path="$sd_install_path/helpers"

if [[ -f $helpers_path/vdf2json.py ]]; then
	echo "FOUND HELPER FILE"
else
	echo "FAILED TO FIND HELPER"
	exit 1
fi

find_library_folder(){
	steam_path=$(python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path')
}
file_picker(){
	echo "OPENING FILE PICKER"
	local path=$(zenity --file-selection --directory 2>/dev/null)
	if [[ -z "$path" ]]; then
		return
	else
		default_steam_path="$path"
		find_library_folder "$default_steam_path"
	fi
	if [[ -z $steam_path ]]; then
		echo "DayZ not found at this path."
		return 1
	else
		echo "$steam_path"
		return 0
	fi
}
main(){
	while true; do
		find_library_folder "$default_steam_path"
		if [[ -z $steam_path ]]; then
			zenity --question --text="DayZ not found or not installed at the chosen path." --ok-label="Choose path manually" --cancel-label="Exit"
			if [[ $? -eq 0 ]]; then
				file_picker
			else
				exit
			fi
		else
			exit 0
		fi
	done
}

main
```

**Comment by dj3hac** [comment]
https://github.com/aclist/dztui/assets/38978989/fceb33af-2c9f-4a58-b099-84bd8ab3350c

This seems to work, no?

**Comment by aclist** [comment]
Alright, I've integrated it verbatim into version `3.2.11`. Let's pray to the DayZ gods that it works now. If not, then the issue is external to the file picker logic and is coming from a different signal.

**Comment by dj3hac** [comment]
*Sad Linux noises*

Same result as before, unable to determine the install path. 

Thank you for all the work you've done so far! You've spent all evening on this, just for me! I have to go to bed soon though, so we can continue this later.

**Comment by aclist** [comment]
Alright, thanks for your patience. Let's revisit this tomorrow or at your convenience. I'll start adding logging to it.

**Comment by aclist** [comment]
I've added some some internal logging that should at least give us a better glimpse of what is happening between the handover from the filepicker dialog to the rest of the script. It writes a log file to `$HOME/.config/dztui/debug.log`. When you have time, please paste or attach the contents here.

**Comment by dj3hac** [comment]
> $HOME/.config/dztui

The directory does not exist.

**Comment by aclist** [comment]
Right. Good point. It wouldn't have been created yet. :crying_cat_face: 

I'll change it to `/tmp/debug.log`. I probably need a coffee.

**Comment by dj3hac** [comment]
> Right. Good point. It wouldn't have been created yet. crying_cat_face
> 
> I'll change it to `/tmp/debug.log`. I probably need a coffee.

Has this been merged? No log file is being created.

Just a quick test. I just got home from work, so I need to play 1 match of Counterstrike as my ritual dictates and then I need to install a new dishwasher :partying_face: 

I should be free later this evening.

**Comment by aclist** [comment]
It is live now.

**Comment by aclist** [comment]
Just to ensure the environment is sane, it would be good if you ran it this way rather than streaming the data to a bash shell:

```
curl <url> > dzgui.sh
chmod +x dzgui.sh
./dzgui.sh
```

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/c1d0ee07-002a-4349-bcd8-4927457c6c05)

**Comment by aclist** [comment]
What is the version number at the top of the script?

**Comment by dj3hac** [comment]
version=3.2.14

**Comment by aclist** [comment]
Try

`ps -p $$`

**Comment by aclist** [comment]
Should have specified about the URL, download the DZGUI script itself and not the bootstrapping script:

```
curl "https://raw.githubusercontent.com/aclist/dztui/dzgui/dzgui.sh" > dzgui.sh
chmod +x dzgui.sh
./dzgui.sh
```

**Comment by aclist** [comment]
I can't find any syntax error in the script itself and it executes without problems on a bash shell. My initial conclusion based on the error output is that your shell is set to execute scripts using fallback POSIX `sh` logic instead of `bash` and is expecting strict encapsulation of the steam path variable. 

You can try the following methods:

1. Invoke the script as `bash dzgui.sh`
2. Change line 1 of the script to `#!/usr/bin/bash`
3. Change line 1 of the script to `#!/usr/bin/env bash`

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/306e987a-c406-4ea7-be60-b716ff9c9871)

![image](https://github.com/aclist/dztui/assets/38978989/e100b2c5-88bd-4c21-b57d-3d1d5419232c)

It did make a log this time :D

**Comment by aclist** [comment]
It's definitely choking on line 233. Either try encapsulating with quotation marks it as follows:

```
OLD:
steam_path=$(python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path')
NEW: 
steam_path="$(python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path')"
```

Or use one of the methods above to expicitly invoke via a bash interpreter.

**Comment by dj3hac** [comment]
> It's definitively choking on line 233. Either try encapsulating it with quotation marks as follows:
> 
> ```
> OLD:
> steam_path=$(python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path')
> NEW: 
> steam_path="$(python3 "$helpers_path/vdf2json.py" -i "$1/steamapps/libraryfolders.vdf" | jq -r '.libraryfolders[]|select(.apps|has("221100")).path')"
> ```
> 
> Or use one of the methods above to expicitly invoke via a bash interpreter.

I just tried all 3 methods as well as changing line 233 no change with any of them it seems.

**Comment by aclist** [comment]
Sorry, looks like we are out of synch here with the timing of messages. 

That log is from a single invocation, right?

It looks like it did actually find the right path after the second iteration:

1. Auto discovery was triggered, found the old $HOME/.local/share/Steam path and set it
2. This did not contain DayZ, so prompted the user to select
3. You selected the Flatpak path
4. It found the DayZ installation

I think I see what to do here, will issue a hotfix in a minute.

**Comment by dj3hac** [comment]
> Sorry, looks like we are out of synch here with the timing of messages.
> 
> That log is from a single invocation, right?
> 
> It looks like it did actually find the right path after the second iteration:
> 
>     1. Auto discovery was triggered, found the old $HOME/.local/share/Steam path and set it
> 
>     2. This did not contain DayZ, so prompted the user to select
> 
>     3. You selected the Flatpak path
> 
>     4. It found the DayZ installation
> 
> 
> I think I see what to do here, will issue a hotfix in a minute.

This is correct

**Comment by aclist** [comment]
Pushed a hotfix, please try as before, thank you.

**Comment by dj3hac** [comment]
Now after I show it where my steam dir is I get stuck in a loop of setting my username and api keys

**Comment by aclist** [comment]
Pushed another fix. So sorry for the staggered approach here, I am trying to get this fixed for you ASAP and go to a meeting, so I'm not testing it as thoroughly as I should be.

**Comment by dj3hac** [comment]
It created the config file!!

![image](https://github.com/aclist/dztui/assets/38978989/39ed8b99-46b6-49cd-93fa-2f799039bee0)

**Comment by aclist** [comment]
OK, wonderful. Sorry this took up more time than necessary.

This was not one isolated bug, but more a complex of problems intertwined with each other, some inside the script, some outside it. I can think of 4-5 issues offhand. In addition to the bug fixed, I'll have to reword some of the documentation and change the method of fetching/installing the script to make it less problematic. This wasn't an issue with your system per se, aside from the problem with the bash interpreter failing at one point.

I'll write up a more detailed post-mortem on this issue later today and itemize what the atomic issues were and how we fixed/can fix them going forward.

**Comment by dj3hac** [comment]
Hmm. the downloading of mods doesn't seem to be working. 

![image](https://github.com/aclist/dztui/assets/38978989/a60902a8-6f28-461e-bccc-c6e68c838194)

I noticed a rather significant bug when family view was enabled; steam (even though it was unlocked) kept asking for the unlock code and the script was unable to proceed.

With family view disabled it seems to just not do anything other than scroll the steam window down for ~30 secs.

**Comment by aclist** [comment]
1. I'm not familiar with family view, how does that work? Is it a pin code to lock out the client?
2. Did you enable auto mod downloads or manual mod downloads?
3. Is this the beta client or the stable client?

**Comment by dj3hac** [comment]
1. It's a 4 digit code that I use for an extra layer of security. It's not necessary for me, so I've disabled it for now.
2. I enabled auto mod downloads from the settings.
3. Steam stable.

**Comment by aclist** [comment]
1. OK, I can look into its functionality and how it interacts with the application anyway.
2. Auto mod downloads in the stable branch is not as robust as in the testing branch. If you want to use auto mod downloads, you should enable the testing branch from inside the application. The updated functionality is being tracked at #60. It is the most experimental feature and has had to undergo considerable refactoring to support changes in the Steam client.
3. Noted

**Comment by aclist** [comment]
> I noticed a rather significant bug when family view was enabled; steam (even though it was unlocked) kept asking for the unlock code and the script was unable to proceed. With family view disabled it seems to just not do anything other than scroll the steam window down for ~30 secs.

I was able to reproduce this as a bug (?) in the Steam client itself. If you enable family view and then turn it on or off (doesn't matter what state it is in), Steam browser protocol commands to open specific parts of the client do not function, or do so sporadically. I can report this upstream as a Steam client bug. Manipulating the Steam client only works if family view is completely disabled via settings.

**Comment by dj3hac** [comment]
I have tried disabling family mode and restarting steam. This cause it to get stuck in another way. Once I select my server and attempt to connect dgz detects the mods and attempts to download them through steam. The dialog says a steam console window will open. By "console window" to you mean terminal, or the steam interface? I get the steam interface. 

The screen will flash between the home tab (Button at the top of your games list) and the last selected game for 30 seconds or so before dgz says it's downloading mods, which it never actually does download the mods. I have to force close this dialog eventually.

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/36ec8588-af25-4690-9d27-a833a96d9ddd)

I see whats going on here haha

**Comment by aclist** [comment]
Steam dev console. Try `steam steam://open/console` and see if it comes up.

This is always going to be one of the weakest links, because we have to do window manipulation and window ID exploration rather than something more programmatic. I'm a bit conflicted about what to do going forward, because what's good for desktop here is not good for Steam Deck, and Beta client necessitates its own logic, so there are a million scenarios to account for.

**Comment by aclist** [comment]
And are you still on Stable branch of DZGUI?

**Comment by dj3hac** [comment]
> And are you still on Stable branch of DZGUI?

No, I've moved to the release candidate.

This did not work for me probably because flatpak: steam steam://open/console

This however got me steam with console: flatpak run com.valvesoftware.Steam -dev

**Comment by dj3hac** [comment]
![image](https://github.com/aclist/dztui/assets/38978989/62f0a37f-a35d-48ab-8fd3-56c1751f36b1)
Shit yeah boye!

It seems I manually had to have the console pre-opened and ready to go. But this is definitely looking good!

**Comment by aclist** [comment]
Alright, we'll need some definitive way of verifying that the user is running Flatpak and then conditionally use that launch command. Does Steam Flatpak have a unique window identifier?

`wmctrl -ilx`

`xprop` and click on the Steam client

Is there a list of launch flags for Steam Flatpak? We need to use a handful of others as well.

**Comment by dj3hac** [comment]
`wmctrl -ilx`
![image](https://github.com/aclist/dztui/assets/38978989/7b8c8363-ee64-4a1e-a195-06864998bc3d)

`xprop`
![image](https://github.com/aclist/dztui/assets/38978989/3483db2d-9cca-4526-a345-31697043b0a9)

**Comment by aclist** [comment]
I was looking through the code, and it seems I added some skeleton code for a manual toggle for Flatpak a long time inside of the advanced options menu but didn't enable it yet. This would allow the user to opt in/out of Flatpak/native Steam rather than trying to autodetect (although we can explore that approach, too).

It should be possible to pass browser protocol commands to Flatpak if using the long form. Does this command work? 

`flatpak run com.valvesoftware.Steam steam steam://open/console`

**Comment by dj3hac** [comment]
Good news! after the mods completed downloading everything went well and I was able to load the mods and connect to the server without any issue!

`flatpak run com.valvesoftware.Steam steam steam://open/console`
![image](https://github.com/aclist/dztui/assets/38978989/0c9a60aa-2f21-4e80-833a-6d6f9e8240ae)

I would say it works, yes.

**Comment by aclist** [comment]
That's good. It's still going to be a bit flaky with targeting the console until I add some updates. But you are on the testing branch now, so those should come more frequently. 

Looking into the legacy code a bit more, currently we have a naive implementation that detects if

- The user has both Steam and Steam Flatpak installed
- The user has one or the other installed

If they have both, we default to Steam. If they only have Flatpak, we default to Flatpak. Main thing missing is enabling the manual toggle so that someone can switch between them. That is just a one line fix. Then you'll get this appear in your config file:

`preferred_client="flatpak run com.valvesoftware.Steam"`

If you don't have vanilla Steam installed, this next update should rectify things out of the box. If you do have concurrent installs, you'll need to toggle it via the advanced options menu.

**Comment by aclist** [comment]
OK, update is live on Testing.

**Comment by dj3hac** [comment]
> The user has both Steam and Steam Flatpak installed

This is likely to be fairly common amongst counterstrike players, csgo doesn't work well on non-flatpak steam. It's why I have flatpak steam installed at all, I eventually moved all my games over because 2 steam clients is annoying. I have uninstalled the RPM version, but it seems like there are still remnants on my system that dzgui is detecting and running with.

I will test the next update momentarily.

**Comment by aclist** [comment]
It makes sense. Unfortunately, Flatpak makes things "easy" for the end-user but is an implementation nightmare due to the syntax and caveats not adhering to anything standard. Seems like we have tamed this beast for now, though. I shouldn't say this, though, because Steam Deck is the real anomaly when it comes to extending support.

Provided you don't have a `steam` binary it should pick up Flatpak only. If it finds both, it will list an option in the menu.

**Comment by aclist** [comment]
>  I'd like to give you some more thorough testing.

Appreciated!

> I would like to test downloading the mods again, it seems that im not subscribed in the workshop. 

If you use manual mod installation, DZGUI just walks you through each workshop page and asks you to click to Subscribe on the mod, then jumps to the next workshop page until they are all done. With automatic mod installation, we are just grabbing the mods and putting them in the DayZ directory directly, so they don't get subscribed to as such. This is a known limitation. However, to counteract this, DZGUI checks your mods for mismatches prior to connecting to a server and will download the applicable mods again if they are out of date. I haven't observed any problems with this logic thus far, as long as the user is aware of the implications. I think only one user reported that they prefer to see mods subscribed to in Steam, and reverted to using manual mode. I think it's strictly a cosmetic issue, but it's something to be aware of going forward in case it has some unintended effect.

If you choose `List installed mods`, you'll see a list of the mod name, symlink, and ID. Navigate to the path shown at the top and `rm -rf` the mod directory name (a string of numbers). Then, navigate to the DayZ installation in your Steam Library and `unlink` the symlink prefixed with an @ symbol that was shown in column 2 of the list.

This is probably not the most elegant approach for now, but I haven't looked into whether deleting mods from inside of DZGUI is reasonable or even desirable. My initial thought process was that DZGUI can install files, but shouldn't destructively try to remove files from the user's system.

**Comment by aclist** [comment]
About having Steam closed before running DZGUI: technically should work, but could introduce problems/race conditions when summoning Steam. I believe we wait for the window to actually spawn, but nonetheless it's preferable to run DZGUI "on top" of Steam for best results.

**Comment by dj3hac** [comment]
> About having Steam closed before running DZGUI: technically should work, but could introduce problems/race conditions when summoning Steam. I believe we wait for the window to actually spawn, but nonetheless it's preferable to run DZGUI "on top" of Steam for best results.

I had multiple instances of steam numerous times during testing because DZGUI would start a new instance rather than working with the existing one.

**Comment by aclist** [comment]
> I had multiple instances of steam numerous times during testing because DZGUI would start a new instance rather than working with the existing one.

This is probably a side effect of Flatpak running each instance as a unique container. While we could check for running instances and block progress, I think this is an anti-pattern, because there could be legitimate reasons someone wants to have multiple clients open outside of the scope of DZGUI.

**Comment by aclist** [comment]
Very nice. Anything else we need to tackle right now? I'll continue hardening the testing branch in advance of a 3.3.0 release and hopefully not break anything else. If you notice anything else out of the ordinary, feel free to post a ticket at any time.

**Comment by dj3hac** [comment]
I think everything is working fine at this point, I will be sure to let you know if anything else comes up! Thank you for your amazing support! I don't think I've ever had a person let alone support division provide such detailed and in-depth support. Thank you so much!

**Comment by aclist** [comment]
Wow, thanks for the donation, that was really unexpected! I didn't even have the link on the repository, but you went out of your way to find it. My humble thanks. I'd definitely like to get your feedback on my next application when I release it.

Post mortem follows, chiefly as an aide-memoire for myself.

The issue presented itself as inability to set the `$default_steam_path` variable during first-time setup when the automatic path discovery logic found files from a concurrent Steam installation not related to the one where the game was installed.

1. The automatic discovery function was not handing control back to the user in the event that the default path did not contain the game. After prompting to set a manual path, the function was not breaking out of the loop and returning to default path discovery. The deeply nested conditional logic was cleaned and arguments are now being passed directly to the function responsible for finding the game library, rather than setting the variables in a different scope.
2. Double-clicking to launch the DZGUI bootstrapping script from a desktop environment file explorer can cause the environment to be set incorrectly, pinning zenity windows to the terminal. This should be documented in the help files so that users do not encounter unexpected behavior that is difficult to diagnose due to its distro-specific nature. In addition, when diagnosing problems, always ask whether the method of invocation was via the terminal. As a final failsafe, we should be able to check the parent PID at runtime to evaluate whether the process was invoked via a proper shell.
3. Streaming the bootstrappings script to a bash shell may cause problems with variable quoting if the system falls back to a POSIX shell. We should more explicitly set the bash interpreter for portability or explicitly advise to run setup without pipes.
4. The prefix used to invoke steam via native Steam or Flatpak commands should be set on all commands leveraging the Steam browser protocol and interpolated. Preferably, move all such commands into a separate function for maintainability. We noted that Flatpak accepts long Steam browser protocol commands when appended verbatim to the `flatpak run com.valvesoftware.Steam` construct. Always use these over obscure launch flags.
5. Family View (FV), an obscure stateful feature, must be disabled to allow automated input. If FV was previously on but set to off, it may still impeded opening the console. Moreover, the console is explicitly blocked when FV state is on. Report this upstream and update the documentation with a caveat.
6. Consider adding functionality for easy removal/management of mods within DZGUI. Possible downsides to this must be weighed over the perceived benefit. Failing this, update the documentation to explicitly describe how to prune mods. Shortly prior to this issue, added a third column ("Dir") to the mods list to allow better identification of mod directories.
7. Due to its containerized nature, Flatpak may allow concurrent instances of Steam to run. This needs additional stress testing. Update the documentation to note that DZGUI should be run "on top of" Steam for best results, rather than allowing Steam to launch after a command is invoked. Waiting on Steam in the background may lead to race conditions due to the timed nature of some commands. This could be further hardened by extending the window ID discovery logic already in place.
8. A final point to note is that a more robust logging framework, rather than an ad-hoc one, should allow for more rapid discovery of issues on a linewise basis. This was previously avoided due to concerns about the negative impact on density of code, but was added in hotfix `3.2.18` and should be used going forward as the primary point of triage for user troubleshooting. Extend this framework into a complete, parameterized log that users can dump into an issue ticket.

**Comment by aclist** [comment]
Hi, just updating to say that all of the fixes we covered have been merged into Stable, so if you want to go back to Stable at this point, you would have no loss in functionality. Testing is going to be getting a steady series of changes around handling the Steam Beta client, so things are liable to break. If you're on the Stable version of the Steam client, though, this shouldn't affect you.

**Comment by dj3hac** [comment]
Thank you so much for reaching out to me! Yes I'm still on the testing branch, I'll move back to stable. I'm staying away from the steam beta for as long as I can. I like my steam skins!On May 22, 2023 08:05, aclist ***@***.***> wrote:
Hi, just updating to say that all of the fixes we covered have been merged into Stable, so if you want to go back to Stable at this point, you would have no loss in functionality. Testing is going to be getting a steady series of changes around handling the Steam Beta client, so things are liable to break. If you're on the Stable version of the Steam client, though, this shouldn't affect you.

**Comment by aclist** [comment]
I think this is the right approach right now, as there is a huge number of regressions in the Steam Beta client that impede a lot of normal functionality of Steam, let alone third party applications.

**Comment by aclist** [comment]
FYI, the Steam UI client beta channel was forced into the Steam stable channel yesterday, so it looks like everyone is on the new UI as of 2023-06-15. If you encounter problems with DZGUI in relation to this change, let me know. The client is still heavily unusable for me, so I can't test all functionality yet.
